### PR TITLE
[MTDSA-6873] - Retrieve a foreign property business' accounting adjustments controller, routes & ITs

### PR DESCRIPTION
**Pull request by DreamcastKnight**

### CODE REVIEW

File: app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+    val event = AuditEvent(
+      auditType = "retrieveBusinessSourceAccountingAdjustments",
+      transactionName = "retrieve-a-foreign-property-business-accounting-adjustments",
+      detail = details
+    )
+
+    auditService.auditEvent(event)
+  }
+}
```
Comment (Mbrookesy) [inline app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:129]: Have we confirmed that we are implementing Audits into this endpoint yet?

File: app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        )
+
+        result
+      }.merge
+    }
+
+  private def errorResult(errorWrapper: ErrorWrapper) = {
+    (errorWrapper.error: @unchecked) match {
+      case BadRequestError | NinoFormatError | BsasIdFormatError => BadRequest(Json.toJson(errorWrapper))
+      case RuleNotSelfEmployment | RuleNoAdjustmentsMade  => Forbidden(Json.toJson(errorWrapper))
```
Comment (Mbrookesy) [inline app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:111]: error should be rule not foreign property

File: test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala [hunk 3]
```diff
@@ -58,4 +71,141 @@ object RetrieveForeignPropertyAdjustmentsFixtures {
...
+       |         "travelCosts": 100.49,
+       |         "costOfServices": 100.49,
+       |         "residentialFinancialCost": 100.49,
+       |         "other": 100.49,
+       |         "consolidatedExpenses": 100.49
+       |      }
+       |   },
+       |	"links": [{
+       |		"href": "/individuals/self-assessment/adjustable-summary/$nino/foreign-property/$bsasId?adjustedStatus=true",
+       |		"method": "GET",
```
Comment (Mbrookesy) [inline test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala:109]: retrieve doesn't have adjusted status as a query parameter for retrieve adjustment

File: test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala [hunk 4]
```diff
@@ -58,4 +71,141 @@ object RetrieveForeignPropertyAdjustmentsFixtures {
...
```
Comment (Mbrookesy) [inline test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala:155]: ditto

File: test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala [hunk 5]
```diff
@@ -58,4 +71,141 @@ object RetrieveForeignPropertyAdjustmentsFixtures {
...
+       |		"href": "/individuals/self-assessment/adjustable-summary/$nino/foreign-property/$bsasId/adjust",
+       |		"method": "GET",
+       |		"rel": "self"
+       |	}]
+       |}
+       |""".stripMargin
+
+  val hateoasResponseForeignPropertyAdjustments: (String, String) => String = (nino: String, bsasId: String) =>
+    s"""
+       |{
```
Comment (Mbrookesy) [inline test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala:121]: why is there duplicate responses?

File: test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+  }
+
+  private val nino = "AA123456A"
+  private val correlationId = "X-123"
+  private val bsasId = "717f3a7a-db8e-11e9-8a34-2a2ae2dbcce5"
+
+  private val request = RetrieveAdjustmentsRequestData(Nino(nino), bsasId)
+  private val requestRawData = RetrieveAdjustmentsRawData(nino, bsasId)
+
+  val testHateoasLinkRetrieveBsas = Link(href = s"/individuals/self-assessment/adjustable-summary/$nino/foreign-property/$bsasId?adjustedStatus=true",
```
Comment (Mbrookesy) [inline test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala:70]: shouldnt have adjusted status query param

File: test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+          }
+        }
+
+        val input = Seq(
+          (NinoFormatError, BAD_REQUEST),
+          (BsasIdFormatError, BAD_REQUEST),
+          (DownstreamError, INTERNAL_SERVER_ERROR),
+          (RuleNoAdjustmentsMade, FORBIDDEN),
+          (NotFoundError, NOT_FOUND),
+          (RuleNotSelfEmployment, FORBIDDEN)
```
Comment (Mbrookesy) [inline test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala:175]: should be rule no foreign property

File: it/v2/endpoints/RetrieveForeignPropertyAdjustmentsControllerISpec.scala [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+      }
+
+      val input = Seq(
+        (BAD_REQUEST, "INVALID_TAXABLE_ENTITY_ID", BAD_REQUEST, NinoFormatError),
+        (BAD_REQUEST, "INVALID_CALCULATION_ID", BAD_REQUEST, BsasIdFormatError),
+        (BAD_REQUEST, "INVALID_RETURN", INTERNAL_SERVER_ERROR, DownstreamError),
+        (UNPROCESSABLE_ENTITY, "UNPROCESSABLE_ENTITY", FORBIDDEN, RuleNoAdjustmentsMade),
+        (NOT_FOUND, "NOT_FOUND", NOT_FOUND, NotFoundError),
+        (INTERNAL_SERVER_ERROR, "SERVER_ERROR", INTERNAL_SERVER_ERROR, DownstreamError),
+        (SERVICE_UNAVAILABLE, "SERVICE_UNAVAILABLE", INTERNAL_SERVER_ERROR, DownstreamError)
```
Comment (Mbrookesy) [inline it/v2/endpoints/RetrieveForeignPropertyAdjustmentsControllerISpec.scala:136]: should try to implement a test for not foreign property

File: app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala (hunk 1 above)
Comment (DreamcastKnight) [reply app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:129]: good question, I'm not sure if we are

Comment (DreamcastKnight) [reply app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:129]: shall I get rid of that then?

File: test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala (hunk 5 above)
Comment (DreamcastKnight) [reply test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala:121]: the duplicate response was to fix the issue I was having in the tests, the integration spec and the unit tests were getting slightly different responses and I wasn't sure what else to do tbh

Comment (DreamcastKnight) [reply test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala:121]: so for example, one had a value for foreignTaxTakenOff, one didn't

it was easier to just make another response

File: app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala (hunk 1 above)
Comment (Mbrookesy) [reply app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:129]: the other one had them removed but feel free to ask james or francis about it

File: test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala (hunk 5 above)
Comment (Mbrookesy) [reply test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala:121]: Hmm it just feels wrong having randomly a bunch of duplicates

Comment (DreamcastKnight) [reply test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala:121]: I mean, they're not exactly duplicates, tbf. they are different

Comment (Mbrookesy) [reply test/v2/fixtures/foreignProperty/RetrieveForeignPropertyAdjustmentsFixtures.scala:121]: Whatever gets them going i guess

File: app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class RetrieveForeignPropertyAdjustmentsController @Inject()(
+                                                              val authService: EnrolmentsAuthService,
+                                                              val lookupService: MtdIdLookupService,
+                                                              requestParser: RetrieveAdjustmentsRequestParser,
+                                                              service: RetrieveForeignPropertyAdjustmentsService,
+                                                              hateoasFactory: HateoasFactory,
+                                                              auditService: AuditService,
```
Comment (Mbrookesy) [inline app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:46]: AuditService is never used

File: it/v2/endpoints/RetrieveForeignPropertyAdjustmentsControllerISpec.scala [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+import support.IntegrationBaseSpec
+import v2.fixtures.foreignProperty.RetrieveForeignPropertyAdjustmentsFixtures._
+import v2.models.errors._
+import v2.stubs.{AuditStub, AuthStub, DesStub, MtdIdLookupStub}
+
+class RetrieveForeignPropertyAdjustmentsControllerISpec extends IntegrationBaseSpec {
+
+  private trait Test {
+    val nino = "AA123456B"
+    val correlationId = "X-123"
```
Comment (Mbrookesy) [inline it/v2/endpoints/RetrieveForeignPropertyAdjustmentsControllerISpec.scala:33]: correlationId is never used

**Review by Mbrookesy — CHANGES_REQUESTED** [review]

File: test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+import play.api.mvc.Result
+import uk.gov.hmrc.domain.Nino
+import uk.gov.hmrc.http.HeaderCarrier
+import v2.models.errors._
+import v2.fixtures.foreignProperty.RetrieveForeignPropertyAdjustmentsFixtures._
+import v2.models.outcomes.ResponseWrapper
+import v2.mocks.hateoas.MockHateoasFactory
+import v2.mocks.requestParsers.MockRetrieveAdjustmentsRequestParser
+import v2.mocks.services._
+import v2.models.audit.{AuditError, AuditEvent, AuditResponse, GenericAuditDetail}
```
Comment (Mbrookesy) [inline test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala:29]: unusued import

File: test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+import v2.models.hateoas.{HateoasWrapper, Link}
+import v2.models.hateoas.Method.GET
+import v2.models.request.{RetrieveAdjustmentsRawData, RetrieveAdjustmentsRequestData}
+import v2.models.response.retrieveBsasAdjustments.foreignProperty.{RetrieveForeignPropertyAdjustmentsHateoasData, RetrieveForeignPropertyAdjustmentsResponse}
```
Comment (Mbrookesy) [inline test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala:33]: unusued import

File: app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+package v2.controllers
+
+import cats.data.EitherT
+import cats.implicits._
+import javax.inject.{Inject, Singleton}
+import play.api.http.MimeTypes
+import play.api.libs.json.Json
+import play.api.mvc.{Action, AnyContent, ControllerComponents}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.audit.http.connector.AuditResult
```
Comment (Mbrookesy) [inline app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:26]: unusued imports

File: app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+import utils.Logging
+import v2.models.errors.ErrorWrapper
+import v2.controllers.requestParsers.RetrieveAdjustmentsRequestParser
+import v2.hateoas.HateoasFactory
+import v2.models.audit.{AuditEvent, AuditResponse, GenericAuditDetail}
```
Comment (Mbrookesy) [inline app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:31]: unusued import

File: app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+import v2.models.errors._
+import v2.models.request.RetrieveAdjustmentsRawData
+import v2.models.response.retrieveBsasAdjustments.foreignProperty.RetrieveForeignPropertyAdjustmentsHateoasData
+import v2.services.{AuditService, EnrolmentsAuthService, MtdIdLookupService, RetrieveForeignPropertyAdjustmentsService}
```
Comment (Mbrookesy) [inline app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:35]: unused import

**Review by Mbrookesy — CHANGES_REQUESTED** [review]

File: it/v2/endpoints/RetrieveForeignPropertyAdjustmentsControllerISpec.scala (hunk 8 above)
Comment (james-work-account) [reply it/v2/endpoints/RetrieveForeignPropertyAdjustmentsControllerISpec.scala:196]: I agree with this ^

File: test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala (hunk 6 above)
Comment (james-work-account) [reply test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala:67]: We don't have query parameters in HATEOAS links IIRC.

**Review by james-work-account — CHANGES_REQUESTED** [review]

File: test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala (hunk 6 above)
Comment (Mbrookesy) [reply test/v2/controllers/RetrieveForeignPropertyAdjustmentsControllerSpec.scala:67]: I'm pretty sure only the submit endpoints have HATEOAS links in one of their links

File: app/v2/models/response/retrieveBsasAdjustments/foreignProperty/IncomeBreakdown.scala [hunk 16]
```diff
@@ -26,7 +26,7 @@ case class IncomeBreakdown(rentIncome: Option[BigDecimal],
 
 object IncomeBreakdown {
   val fhlReads: Reads[IncomeBreakdown] = (
-    (JsPath \ "totalRentsReceived").readNullable[BigDecimal] and
+    (JsPath \ "rentReceived").readNullable[BigDecimal] and
```
Comment (james-work-account) [inline app/v2/models/response/retrieveBsasAdjustments/foreignProperty/IncomeBreakdown.scala:29]: What is this change for?

Comment (james-work-account) [reply app/v2/models/response/retrieveBsasAdjustments/foreignProperty/IncomeBreakdown.scala:29]: The spec says `adjustments.income.totalRentsReceived` is mapped to `adjustments.incomes.rentIncome` - this was correct before. Please change it back (or explain the reason behind this change)

File: test/v2/models/response/retrieveBsasAdjustments/foreignProperty/BsasDetailSpec.scala [hunk 17]
```diff
@@ -27,7 +27,7 @@ class BsasDetailSpec extends UnitSpec with JsonErrorValidators {
     """{
       | "adjustments" : {
       |    "income": {
-      |      "totalRentsReceived": 100.49
+      |      "rentReceived": 100.49
```
Comment (james-work-account) [inline test/v2/models/response/retrieveBsasAdjustments/foreignProperty/BsasDetailSpec.scala:30]: See above - what is this change for?

File: test/v2/models/response/retrieveBsasAdjustments/foreignProperty/IncomeBreakdownSpec.scala [hunk 18]
```diff
@@ -25,7 +25,7 @@ class IncomeBreakdownSpec extends UnitSpec with JsonErrorValidators {
 
   val fhlDesJson: JsValue = Json.parse(
     """{
-      |         "totalRentsReceived": 100.49
+      |         "rentReceived": 100.49
```
Comment (james-work-account) [inline test/v2/models/response/retrieveBsasAdjustments/foreignProperty/IncomeBreakdownSpec.scala:28]: See above - what is this change for?

File: test/v2/models/response/retrieveBsasAdjustments/foreignProperty/RetrieveForeignPropertyAdjustmentsResponseSpec.scala [hunk 19]
```diff
@@ -39,7 +39,7 @@ class RetrieveForeignPropertyAdjustmentsResponseSpec extends UnitSpec with JsonE
       | },
       | "adjustments" : {
       |    "income": {
-      |      "totalRentsReceived": 100.49
+      |      "rentReceived": 100.49
```
Comment (james-work-account) [inline test/v2/models/response/retrieveBsasAdjustments/foreignProperty/RetrieveForeignPropertyAdjustmentsResponseSpec.scala:42]: See above - what is this change for?

**Review by james-work-account — CHANGES_REQUESTED** [review]

File: app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+import cats.data.EitherT
+import cats.implicits._
+import javax.inject.{Inject, Singleton}
+import play.api.http.MimeTypes
+import play.api.libs.json.Json
+import play.api.mvc.{Action, AnyContent, ControllerComponents}
+import utils.Logging
+import v2.controllers.requestParsers.RetrieveAdjustmentsRequestParser
+import v2.hateoas.HateoasFactory
+import v2.models.errors.{ErrorWrapper, _}
```
Comment (Nboaram) [inline app/v2/controllers/RetrieveForeignPropertyAdjustmentsController.scala:28]: why specify one error and then use ._?

**Review by Nboaram — CHANGES_REQUESTED** [review]

**Review by Nboaram — APPROVED** [review]

**Pull request merged**